src/jit/codegenxarch.cpp

```diff
@@ -3921,41 +3999,22 @@ void CodeGen::genStructPutArgUnroll(GenTreePutArgStk* putArgNode, unsigned baseV
     // Fill the remainder (15 bytes or less) if there's one.
     if ((size & 0xf) != 0)
     {
-        // Grab the integer temp register to emit the remaining loads and stores.
-        regNumber tmpReg = genRegNumFromMask(putArgNode->gtRsvdRegs & RBM_ALLINT);
-        assert(genIsValidIntReg(tmpReg));
-
-        if ((size & 8) != 0)
-        {
-            genCodeForLoadOffset(INS_mov, EA_8BYTE, tmpReg, src->gtOp.gtOp1, offset);
-
-            emit->emitIns_S_R(INS_mov, EA_8BYTE, tmpReg, baseVarNum, putArgOffset + offset);
-
-            offset += 8;
-        }
-
-        if ((size & 4) != 0)
-        {
-            genCodeForLoadOffset(INS_mov, EA_4BYTE, tmpReg, src->gtOp.gtOp1, offset);
-
-            emit->emitIns_S_R(INS_mov, EA_4BYTE, tmpReg, baseVarNum, putArgOffset + offset);
-
-            offset += 4;
-        }
-
-        if ((size & 2) != 0)
+#ifdef _TARGET_X86_
+        if (m_pushStkArg)
         {
-            genCodeForLoadOffset(INS_mov, EA_2BYTE, tmpReg, src->gtOp.gtOp1, offset);
-
-            emit->emitIns_S_R(INS_mov, EA_2BYTE, tmpReg, baseVarNum, putArgOffset + offset);
-
-            offset += 2;
+            offset = (size & 0xf);
+            offset -= genMove8IfNeeded(size, longTmpReg, src->gtOp.gtOp1, offset - 8);
+            offset -= genMove4IfNeeded(size, intTmpReg, src->gtOp.gtOp1, offset - 4);
+            offset -= genMove2IfNeeded(size, intTmpReg, src->gtOp.gtOp1, offset - 2);
+            offset -= genMove1IfNeeded(size, intTmpReg, src->gtOp.gtOp1, offset - 1);
```

Comment:
Actually, looking at this, the last argument is ignored in the `m_pushStkArg` case (and should probably be 0 for consistency, since when we're pushing we always push at offset 0).

---
In reply to: [79525935](https://github.com/dotnet/coreclr/pull/7252#discussion_r79525935) [](ancestors = 79525935)